pkg/controllers/management/auth/globalroles/globalrole_handler_test.go

```diff
@@ -0,0 +1,686 @@
+package globalroles
+
+import (
+	"fmt"
+	"testing"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	normanFakes "github.com/rancher/rancher/pkg/generated/norman/core/v1/fakes"
+	normanv1 "github.com/rancher/rancher/pkg/generated/norman/rbac.authorization.k8s.io/v1"
+	rbacFakes "github.com/rancher/rancher/pkg/generated/norman/rbac.authorization.k8s.io/v1/fakes"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	readPodPolicyRule = rbacv1.PolicyRule{
+		Verbs:     []string{"get", "list", "watch"},
+		APIGroups: []string{""},
+		Resources: []string{"pods"},
+	}
+
+	readConfigPolicyRule = rbacv1.PolicyRule{
+		Verbs:     []string{"get", "list", "watch"},
+		APIGroups: []string{""},
+		Resources: []string{"configmaps"},
+	}
+	adminPodPolicyRule = rbacv1.PolicyRule{
+		Verbs:     []string{"*"},
+		APIGroups: []string{""},
+		Resources: []string{"pod"},
+	}
+
+	namespacedRulesGR = v3.GlobalRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "namespacedRulesGR",
+			UID:  "00000000",
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "fake-kind",
+			APIVersion: "fake-version",
+		},
+		NamespacedRules: map[string][]rbacv1.PolicyRule{
+			"namespace1": {
+				readPodPolicyRule,
+				readConfigPolicyRule,
+			},
+			"namespace2": {
+				adminPodPolicyRule,
+				readPodPolicyRule,
+			},
+		},
+	}
+	namespacedRulesOwnerRef = metav1.OwnerReference{
+		APIVersion: namespacedRulesGR.APIVersion,
+		Kind:       namespacedRulesGR.Kind,
+		Name:       namespacedRulesGR.Name,
+		UID:        namespacedRulesGR.UID,
+	}
+
+	updatedNamespacedRulesGR = v3.GlobalRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "namespacedRulesGR",
+		},
+		NamespacedRules: map[string][]rbacv1.PolicyRule{
+			"namespace1": {
+				readPodPolicyRule,
+			},
+			"namespace2": {
+				adminPodPolicyRule,
+				readPodPolicyRule,
+				readConfigPolicyRule,
+			},
+		},
+	}
+)
+
+type grTestStateChanges struct {
+	t            *testing.T
+	createdRoles map[string]*rbacv1.Role
+	deletedRoles []string
```

Comment:
I think it would be safer to have deletedRoles be a map instead of a list. My reasoning is that since the reconciler operates on a map[namespace][]rules, the order of the loop is non deterministic.

So if you were adding tests that deletes multiple roles in different namespaces, you could have one day 
- delete from ns1
- delete from ns2

which would work, but then once in a while you'd end up with
- delete from ns2
- delete from ns1
failing the test (because you'd assume deletedRoles[0] is always from ns1), leading to flaky test.

Note that this isn't the case now because only a single delete is ever tested, so it's more of a "let's future proof this" kinda thing.